eth/protocols/eth/peer.go

```diff
@@ -331,63 +343,88 @@ func (p *Peer) SendBlockHeaders(headers []*types.Header) error {
 	return p2p.Send(p.rw, BlockHeadersMsg, BlockHeadersPacket(headers))
 }
 
-// SendBlockBodies sends a batch of block contents to the remote peer.
-func (p *Peer) SendBlockBodies(bodies []*BlockBody) error {
-	return p2p.Send(p.rw, BlockBodiesMsg, BlockBodiesPacket(bodies))
+func (p *Peer) ReplyBlockHeaders(id uint64, headers BlockHeadersPacket) error {
+	return p2p.Send(p.rw, BlockHeadersMsg, BlockHeadersPacket66{id, headers})
 }
 
 // SendBlockBodiesRLP sends a batch of block contents to the remote peer from
 // an already RLP encoded format.
-func (p *Peer) SendBlockBodiesRLP(bodies []rlp.RawValue) error {
+func (p *Peer) SendBlockBodiesRLP(bodies BlockBodiesRLPPacket) error {
 	return p2p.Send(p.rw, BlockBodiesMsg, bodies) // Not packed into BlockBodiesPacket to avoid RLP decoding
 }
 
+func (p *Peer) ReplyBlockBodiesRLP(id uint64, bodies BlockBodiesRLPPacket) error {
+	return p2p.Send(p.rw, BlockBodiesMsg, BlockBodiesRLPPacket66{id, bodies}) // Not packed into BlockBodiesPacket to avoid RLP decoding
+}
+
 // SendNodeDataRLP sends a batch of arbitrary internal data, corresponding to the
 // hashes requested.
-func (p *Peer) SendNodeData(data [][]byte) error {
-	return p2p.Send(p.rw, NodeDataMsg, NodeDataPacket(data))
+func (p *Peer) SendNodeData(data NodeDataPacket) error {
+	return p2p.Send(p.rw, NodeDataMsg, data)
+}
+
+// ReplyNodeData is the eth/66 response to GetNodeData
+func (p *Peer) ReplyNodeData(id uint64, data NodeDataPacket) error {
+	return p2p.Send(p.rw, NodeDataMsg, NodeDataPacket66{id, data})
 }
 
 // SendReceiptsRLP sends a batch of transaction receipts, corresponding to the
 // ones requested from an already RLP encoded format.
-func (p *Peer) SendReceiptsRLP(receipts []rlp.RawValue) error {
+func (p *Peer) SendReceiptsRLP(receipts ReceiptsRLPPacket) error {
 	return p2p.Send(p.rw, ReceiptsMsg, receipts) // Not packed into ReceiptsPacket to avoid RLP decoding
 }
 
+// ReplyReceiptsRLP is the eth/66 response to GetReceipts
+func (p *Peer) ReplyReceiptsRLP(id uint64, receipts ReceiptsRLPPacket) error {
+	return p2p.Send(p.rw, ReceiptsMsg, ReceiptsRLPPacket66{id, receipts})
+}
+
 // RequestOneHeader is a wrapper around the header query functions to fetch a
 // single header. It is used solely by the fetcher.
 func (p *Peer) RequestOneHeader(hash common.Hash) error {
 	p.Log().Debug("Fetching single header", "hash", hash)
-	return p2p.Send(p.rw, GetBlockHeadersMsg, &GetBlockHeadersPacket{
+	query := &GetBlockHeadersPacket{
```

Comment:
This was wrong (fix upcoming) -- it's a pointer here, and gets doubly referenced on `eth65` on L395